### PR TITLE
Laboratory work #3, Blyudova Vasilisa - 20FPL3

### PR DESCRIPTION
**Pull request by Vasilisa-Blyudova**

### CODE REVIEW

File: lab_3/main.py [hunk 1]
```diff
@@ -20,6 +21,31 @@ def tokenize_by_sentence(text: str) -> tuple:
          (('_', 'h', 'e', '_'), ('_', 'i', 's', '_'), ('_', 'h', 'a', 'p', 'p', 'y', '_'))
          )
     """
+    if not isinstance(text, str):
+        return ()
+    patterns = ['ö', 'ü', 'ä', 'ß']
+    replacements = ['oe', 'ue', 'ae', 'ss']
```
Comment (WhiteJaeger) [inline lab_3/main.py:27]: If you are not going to change them, it is better to use `tuple`s instead of `list`s.

File: lab_3/main.py [hunk 2]
```diff
@@ -20,6 +21,31 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+    lst_of_sentences = [sentence.lower() for sentence in sentences]
+    new_sentence_tuple = []
+    for sentence in lst_of_sentences:
+        sentence = sentence.split(' ')
+        new_words_lst = []
+        for word in sentence:
+            for pattern, replacement in zip(patterns, replacements):
+                word.replace(pattern, replacement)
+            lst_of_letters = [letter for letter in word if letter.isalpha()]
+            if len(lst_of_letters) != 0:
```
Comment (WhiteJaeger) [inline lab_3/main.py:40]: It can be done without comparison.

File: lab_3/main.py [hunk 3]
```diff
@@ -20,6 +21,31 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+        return ()
+    patterns = ['ö', 'ü', 'ä', 'ß']
+    replacements = ['oe', 'ue', 'ae', 'ss']
+    sentences = re.split(r"[.!?] ?", text)
+    if '' in sentences:
+        sentences.remove('')
+    lst_of_sentences = [sentence.lower() for sentence in sentences]
+    new_sentence_tuple = []
+    for sentence in lst_of_sentences:
+        sentence = sentence.split(' ')
```
Comment (WhiteJaeger) [inline lab_3/main.py:34]: Space is the default delimiter, no need to explicitly set it.

File: lab_3/main.py [hunk 4]
```diff
@@ -20,6 +21,31 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+        for word in sentence:
+            for pattern, replacement in zip(patterns, replacements):
+                word.replace(pattern, replacement)
+            lst_of_letters = [letter for letter in word if letter.isalpha()]
+            if len(lst_of_letters) != 0:
+                lst_of_letters.insert(0, '_')
+                lst_of_letters.append('_')
+                new_words_lst.append(tuple(lst_of_letters))
+                print(new_words_lst)
+            if len(new_words_lst) == 0:
```
Comment (WhiteJaeger) [inline lab_3/main.py:45]: Can be done without comparison.

File: lab_3/main.py [hunk 5]
```diff
@@ -20,6 +21,31 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+                return ()
+        new_sentence_tuple.append(tuple(new_words_lst))
+    return tuple(new_sentence_tuple)
     pass
```
Comment (WhiteJaeger) [inline lab_3/main.py:49]: Remove as it is unused.

File: lab_3/main.py [hunk 6]
```diff
@@ -29,23 +55,32 @@ class LetterStorage:
     Stores and manages letters
     """
 
-    def __init__(self):
+    def __init__(self, count=0):
```
Comment (WhiteJaeger) [inline lab_3/main.py:58]: Please, do not alter the method definitions. Define the `count` inside the constructor and set the default value.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]
Good start yet some comments.

File: lab_3/main.py (hunk 1 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py]: done

File: lab_3/main.py (hunk 2 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py]: done

File: lab_3/main.py (hunk 3 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py]: done

File: lab_3/main.py (hunk 4 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py]: done

File: lab_3/main.py (hunk 5 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py:48]: Do I understand correctly that it is necessary to remove "print" on line 44?

File: lab_3/main.py (hunk 6 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py]: done

File: lab_3/main.py (hunk 5 above)
Comment (WhiteJaeger) [reply lab_3/main.py:48]: I meant the line with `pass`.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]
Please, remove the lines with `pass`. And do not forget to change the `target_score.txt` if you wish the score above 6.

File: lab_3/main.py (hunk 4 above)
Comment (WhiteJaeger) [reply lab_3/main.py:45]: I do not see the change.

File: lab_3/main.py [hunk 7]
```diff
@@ -84,16 +143,23 @@ def decode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
...
     """
     Stores and manages ngrams
     """
-    
     def __init__(self, n: int, letter_storage: LetterStorage):
+        self.storage = LetterStorage()
+        self.size = n
+        self.n_grams = []
+        self.n_gram_frequencies = {}
         pass
```
Comment (WhiteJaeger) [inline lab_3/main.py:163]: Remove the `pass` statements.

File: lab_3/main.py [hunk 8]
```diff
@@ -222,7 +331,24 @@ def save(self, name: str) -> int:
...
+                for key, value in trie.n_gram_frequencies.items():
+                    for letter_id in key:
+                        new_key += self.storage.get_letter_by_id(letter_id)
+                    freq_dict[new_key] = value
+                    new_key = ''
+            profile['freq'] = freq_dict
+            profile['n_words'] = self.n_words
+            profile['name'] = self.language
+            json_string = json.dumps(profile)
+            file.write(json_string)
```
Comment (WhiteJaeger) [inline lab_3/main.py:350]: There is the `.dump` method in the `json` package which combines these two lines.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_3/main.py (hunk 7 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py]: done

File: lab_3/main.py (hunk 8 above)
Comment (Vasilisa-Blyudova) [reply lab_3/main.py]: done

**Pull request closed without merging**